models/model_upload/multimodal_models/vllm-miniCPM-2.6/1/model.py

```diff
@@ -0,0 +1,304 @@
+        base_url=openai_api_base,
+    )
+    self.server_started_event = threading.Event()
+
+    self.python_executable = sys.executable
+
+    # Start the VLLM server in a separate thread
+    vllm_server_thread = threading.Thread(target=self.vllm_openai_server)
+    vllm_server_thread.start()
+
+    # Wait for the server to be ready
+    self.server_started_event.wait()
+
+    models = self.client.models.list()
+    self.model = models.data[0].id
+
+  def vllm_openai_server(self):
+    os.path.join(os.path.dirname(__file__))
+
+    # if checkpoints section is in config.yaml file then checkpoints will be downloaded at this path during model upload time.
+    checkpoints = os.path.join(os.path.dirname(__file__), "checkpoints")
+    process = subprocess.Popen(
+        [
+            self.python_executable,
+            '-m',
+            'vllm.entrypoints.openai.api_server',
+            '--model',
+            checkpoints,
+            '--dtype',
+            'float16',
+            '--tensor-parallel-size',
+            '1',
+            '--quantization',
+            'awq',
+            '--gpu-memory-utilization',
+            '0.8',
+            '--port',
+            str(self.port),
+            '--host',
+            'localhost',
+            '--max-model-len',
+            '2048',
+            "--trust-remote-code",
+        ],
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        text=True,
+    )
+
+    # Check the server output to confirm it's running
+    for line in process.stderr:
+      line = line.strip()
+      logger.info(line)
+      if "Uvicorn running on http://localhost:" in line:
+        self.server_started_event.set()
+        logger.info(f"vLLM Server started at http://localhost:{self.port}")
+        break
+
+  def predict(self, request: service_pb2.PostModelOutputsRequest
+             ) -> Iterator[service_pb2.MultiOutputResponse]:
+    """This is the method that will be called when the runner is run. It takes in an input and
+    returns an output.
+    """
+
+    # TODO: Could cache the model and this conversion if the hash is the same.
+    model = request.model
+    output_info = None
+    if request.model.model_version.id != "":
+      output_info = json_format.MessageToDict(
+          model.model_version.output_info, preserving_proto_field_name=True)
+
+    outputs = []
+    # TODO: parallelize this over inputs in a single request.
+    for inp in request.inputs:
+      output = resources_pb2.Output()
+
+      data = inp.data
+
+      # Optional use of output_info
+      inference_params = {}
+      if "params" in output_info:
+        inference_params = output_info["params"]
+
+      messages = []
+      temperature = inference_params.get("temperature", 0.7)
+      max_tokens = inference_params.get("max_tokens", 100)
+      top_p = inference_params.get("top_p", 1.0)
+
+      prompt = None
+      image_url = None
+      image_bytes = None
+      if data.text.raw != "":
```

Comment:
shouldn't we be using the new data.parts list to get text and image passed to these models so we start converting all models to that new format so users get used to it and can support multiple text and images?

Reply:
should be completely depreciate using `data. text ` and `data.image` during model upload in platform and in all the model examples in this repo or should be support both `data.image` and `data.parts `?

Reply:
@zeiler 
if the model supports single `image` and single `text` as input then should be use `data.parts` or `data.image` and `data.text`?

Reply:
i like the idea of getting people used to data.parts so that we have one way to support all multimodal models. Curious what you and @phatvo9 and @ackizilkale think?

Reply:
Agreed, I think we should maintain consistency in how we handle protos across all models. It’s better to standardize on using `data.parts` for all models, not just for multimodal models.

The same question applies to model output as well. Should we use `output.data.parts` there too? Many models support generating multiple answers for a single input, so with `output.data.parts`, we can easily support it.  What do you think?